modules/cas_cache/utils/utils_mpool.c

```diff
@@ -3,34 +3,61 @@
  * SPDX-License-Identifier: BSD-3-Clause-Clear
  */
 
-#include "ocf_env.h"
 #include "utils_mpool.h"
+#include "ocf_env.h"
+
+struct env_mpool {
+	size_t (*size_of)(uint32_t);
+		/*!< Getter for allocation size for given item count */
+
+	env_allocator *allocator[env_mpool_max];
+		/*!< OS handle to memory pool */
+
+	bool fallback;
+		/*!< Should mpool fallback to vmalloc */
 
+	int flags;
+		/*!< Allocation flags */
+};
 
-struct cas_mpool *cas_mpool_create(uint32_t hdr_size, uint32_t size, int flags,
-		int mpool_max, const char *name_perfix)
+struct env_mpool *env_mpool_create(size_t (*size_of)(uint32_t), int flags,
+		int mpool_max, bool fallback,
+		const struct env_mpool_init_limit *limits,
+		const char *name_perfix)
 {
 	uint32_t i;
-	char name[ALLOCATOR_NAME_MAX] = { '\0' };
+	char name[MPOOL_ALLOCATOR_NAME_MAX] = { '\0' };
 	int result;
-	struct cas_mpool *mpool;
+	struct env_mpool *mpool;
+	int rpool_limit = -1;
+	int limit_iter = 0;
+	size_t size;
 
 	mpool = env_zalloc(sizeof(*mpool), ENV_MEM_NORMAL);
 	if (!mpool)
 		return NULL;
 
-	mpool->item_size = size;
-	mpool->hdr_size = hdr_size;
 	mpool->flags = flags;
+	mpool->fallback = fallback;
+	mpool->size_of = size_of;
 
-	for (i = 0; i < min(cas_mpool_max, mpool_max + 1); i++) {
+	for (i = 0; i < min(env_mpool_max, mpool_max + 1); i++) {
 		result = snprintf(name, sizeof(name), "%s_%u", name_perfix,
 				(1 << i));
 		if (result < 0 || result >= sizeof(name))
 			goto err;
 
-		mpool->allocator[i] = env_allocator_create(
-				hdr_size + (size * (1 << i)), name);
+		size = size_of(1 << i);
+
+		if (limits &&
+			limits[limit_iter].size &&
+			limits[limit_iter].size <= size) {
+			rpool_limit = limits[limit_iter].rpool_limit;
+			limit_iter++;
+		}
```

Comment:
in "else" case you are stuck with rpool_limit  from previous loop iteration

Reply:
That was the intention - this way the limit will be applied to every allocation over the threshold until next threshold is exceeded